core/src/main/java/com/venia/core/models/commerce/services/ClearFullDispatcherCacheBasedOnCategory.java

```diff
@@ -0,0 +1,248 @@
+            return Collections.emptyList();
+        }
+
+        return categories.stream()
+                .map(category -> {
+                    Map<String, Object> map = new HashMap<>();
+                    map.put("uid", category.getUid());
+                    map.put(LEVEL_KEY, category.getLevel());
+                    return map;
+                })
+                .collect(Collectors.toList());
+    }
+
+    protected Query getGraphqlResponseData(MagentoGraphqlClient client, String query) {
+        if (client == null || query == null) {
+            return null;
+        }
+
+        GraphqlResponse<com.adobe.cq.commerce.magento.graphql.Query, Error> response = client.execute(query);
+
+        if (response == null || (response.getErrors() != null && !response.getErrors().isEmpty()) || response.getData() == null) {
+            return null;
+        }
+
+        return response.getData();
+    }
+
+    /**
+     * Builds the GraphQL query for fetching category information.
+     *
+     * @param categoryUids Array of category UIDs to query
+     * @return The GraphQL query string
+     */
+    private String buildCategoryQuery(String[] categoryUids) {
+        if (categoryUids == null || categoryUids.length == 0) {
+            LOGGER.warn("Invalid category UIDs for building query");
+            return null;
+        }
+
+        try {
+            CategoryFilterInput filter = new CategoryFilterInput();
+            FilterEqualTypeInput identifiersFilter = new FilterEqualTypeInput().setIn(Arrays.asList(categoryUids));
+            filter.setCategoryUid(identifiersFilter);
+            QueryQuery.CategoryListArgumentsDefinition searchArgs = s -> s.filters(filter);
+
+            CategoryTreeQueryDefinition queryArgs = q -> q.uid().level();
+
+            return Operations.query(query -> query
+                    .categoryList(searchArgs, queryArgs)).toString();
+        } catch (Exception e) {
+            LOGGER.error("Error building category query", e);
+            return null;
+        }
+    }
+
+    /**
+     * Validates if the provided category UIDs array is valid for processing.
+     *
+     * @param categoryUids Array of category UIDs to validate
+     * @return true if category UIDs are valid, false otherwise
+     */
+    private boolean isValidCategoryUids(String[] categoryUids) {
+        if (categoryUids == null || categoryUids.length == 0) {
+            LOGGER.warn("No category UIDs provided for cache invalidation");
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Retrieves the navigation structure depth from the header configuration.
+     *
+     * @param resourceResolver The resource resolver
+     * @return The navigation structure depth or null if not found
+     */
+    private Integer getNavigationStructureDepth(ResourceResolver resourceResolver) {
+        if (resourceResolver == null) {
```

Comment:
This should never be null either. You should guarantee that out of the box, and add it to the documentation.